File: arduinobot_task_planner/arduinobot_task_planner/scene_manager.py
import rclpy
from rclpy.node import Node
from moveit_msgs.msg import CollisionObject
from shape_msgs.msg import SolidPrimitive
from geometry_msgs.msg import Pose


class SceneManager(Node):
    def __init__(self):
        super().__init__('scene_manager')

        self.collision_pub = self.create_publisher(CollisionObject, '/collision_object', 10)

        self.publish_count = 0
        self.max_publishes = 5

        self.get_logger().info('Waiting for MoveIt2 — 10 seconds to start...')

        # Wait 10 seconds before starting
        self.create_timer(10.0, self.start_publishing)

    def start_publishing(self):
        # Publish every 2 seconds, 5 times
        self.publish_timer = self.create_timer(2.0, self.publish_objects)

    def publish_objects(self):
        if self.publish_count >= self.max_publishes:
            self.publish_timer.cancel()
            self.get_logger().info('Publishing completed')
            return

        self.publish_count += 1
        self.get_logger().info(
            f'Publishing {self.publish_count}/{self.max_publishes}'
        )
        self.add_table()
        self.add_box()

    def add_table(self):
        obj = CollisionObject()
        obj.id = 'table'
        obj.header.frame_id = 'world'
        obj.header.stamp = self.get_clock().now().to_msg()

        shape = SolidPrimitive()
        shape.type = SolidPrimitive.BOX
        shape.dimensions = [0.6, 0.6, 0.05]

        pose = Pose()
        pose.position.x = 0.0
        pose.position.y = 0.0
        pose.position.z = -0.05
        pose.orientation.w = 1.0

        obj.primitives = [shape]
        obj.primitive_poses = [pose]
        obj.operation = CollisionObject.ADD

        self.collision_pub.publish(obj)

    def add_box(self):
        obj = CollisionObject()
        obj.id = 'obstacle'
        obj.header.frame_id = 'world'
        obj.header.stamp = self.get_clock().now().to_msg()

        shape = SolidPrimitive()
        shape.type = SolidPrimitive.BOX
        shape.dimensions = [0.05, 0.05, 0.3]

        pose = Pose()
        pose.position.x = 0.15
        pose.position.y = 0.15
        pose.position.z = 0.15
        pose.orientation.w = 1.0

        obj.primitives = [shape]
        obj.primitive_poses = [pose]
        obj.operation = CollisionObject.ADD

        self.collision_pub.publish(obj)


def main(args=None):
    rclpy.init(args=args)
    node = SceneManager()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: arduinobot_task_planner_cpp/src/scene_manager.cpp
#include <memory>
#include <vector>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "moveit_msgs/msg/collision_object.hpp"
#include "shape_msgs/msg/solid_primitive.hpp"
#include "geometry_msgs/msg/pose.hpp"

using moveit_msgs::msg::CollisionObject;

class SceneManager : public rclcpp::Node
{
public:
    SceneManager()
        : Node("scene_manager"),
          publish_count_(0),
          max_publishes_(5)
    {
        collision_pub_ = this->create_publisher<CollisionObject>(
            "/collision_object", 10);

        RCLCPP_INFO(this->get_logger(),
                    "Waiting for MoveIt2 — 10 seconds to start...");

        start_timer_ = this->create_wall_timer(
            std::chrono::seconds(10),
            std::bind(&SceneManager::startPublishing, this));
    }

private:

    // ==========================================================
    // START PERIODIC PUBLISHING
    // ==========================================================
    void startPublishing()
    {
        start_timer_->cancel();

        publish_timer_ = this->create_wall_timer(
            std::chrono::seconds(2),
            std::bind(&SceneManager::publishObjects, this));
    }

    // ==========================================================
    // MAIN LOOP
    // ==========================================================
    void publishObjects()
    {
        if (publish_count_ >= max_publishes_)
        {
            publish_timer_->cancel();
            RCLCPP_INFO(this->get_logger(), "Publishing completed");
            return;
        }

        publish_count_++;

        RCLCPP_INFO(this->get_logger(),
                    "Publishing %d/%d",
                    publish_count_,
                    max_publishes_);

        addTable();
        addBox();
    }

    // ==========================================================
    // TABLE OBJECT
    // ==========================================================
    void addTable()
    {
        CollisionObject obj;
        obj.id = "table";
        obj.header.frame_id = "world";
        obj.header.stamp = this->now();

        shape_msgs::msg::SolidPrimitive shape;
        shape.type = shape_msgs::msg::SolidPrimitive::BOX;
        shape.dimensions = {0.6, 0.6, 0.05};

        geometry_msgs::msg::Pose pose;
        pose.position.x = 0.0;
        pose.position.y = 0.0;
        pose.position.z = -0.05;
        pose.orientation.w = 1.0;

        obj.primitives.push_back(shape);
        obj.primitive_poses.push_back(pose);
        obj.operation = CollisionObject::ADD;

        collision_pub_->publish(obj);
    }

    // ==========================================================
    // OBSTACLE BOX
    // ==========================================================
    void addBox()
    {
        CollisionObject obj;
        obj.id = "obstacle";
        obj.header.frame_id = "world";
        obj.header.stamp = this->now();

        shape_msgs::msg::SolidPrimitive shape;
        shape.type = shape_msgs::msg::SolidPrimitive::BOX;
        shape.dimensions = {0.05, 0.05, 0.3};

        geometry_msgs::msg::Pose pose;
        pose.position.x = 0.15;
        pose.position.y = 0.15;
        pose.position.z = 0.15;
        pose.orientation.w = 1.0;

        obj.primitives.push_back(shape);
        obj.primitive_poses.push_back(pose);
        obj.operation = CollisionObject::ADD;

        collision_pub_->publish(obj);
    }

private:

    // Publisher
    rclcpp::Publisher<CollisionObject>::SharedPtr collision_pub_;

    // Timers
    rclcpp::TimerBase::SharedPtr start_timer_;
    rclcpp::TimerBase::SharedPtr publish_timer_;

    // State
    int publish_count_;
    int max_publishes_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);

    auto node = std::make_shared<SceneManager>();

    rclcpp::spin(node);

    rclcpp::shutdown();
    return 0;
}